### PR TITLE
Update az snp / tdx vtpm dependency to 0.5

### PR DESCRIPTION
**Pull request by surajssd**

### CODE REVIEW

**Comment by fitzthum** [comment]
I think you'll want to do `cargo update az-snp-vtpm` and commit the Cargo.lock

**Comment by surajssd** [comment]
I think I will wait until https://github.com/confidential-containers/guest-components/pull/436 is merged.

**Comment by fitzthum** [comment]
The stable check seems to have a legit failure. Maybe there were changes to the crate?

**Comment by surajssd** [comment]
> The stable check seems to have a legit failure. Maybe there were changes to the crate?

Yes that's right. I am on it.

**Comment by surajssd** [comment]
cc: @mkulke for review of the updated tests.

**Comment by surajssd** [comment]
Can someone authorize the test runs?

File: attestation-service/verifier/src/az_snp_vtpm/mod.rs [hunk 1]
```diff
@@ -134,14 +134,13 @@ fn verify_snp_report(
...
-    const SIGNATURE: &[u8; 256] = include_bytes!("../../test_data/az-snp-vtpm/tpm-quote.sig");
-    const MESSAGE: &[u8; 122] = include_bytes!("../../test_data/az-snp-vtpm/tpm-quote.msg");
+    const REPORT: &[u8; 3848] = include_bytes!("../../test_data/az-snp-vtpm/hcl-report.bin");
+    const QUOTE: &[u8; 1362] = include_bytes!("../../test_data/az-snp-vtpm/quote.bin");
     const REPORT_DATA: &[u8] = "challenge".as_bytes();
 
     #[test]
     fn test_verify_snp_report() {
-        let hcl_report = HclReport::new(REPORT.to_vec()).unwrap();
+        let hcl_report = HclReport::new(bincode::deserialize(REPORT).unwrap()).unwrap();
```
Comment (mkulke) [inline attestation-service/verifier/src/az_snp_vtpm/mod.rs:143]: the HCL report should be just bytes, why do we need to deserialize `REPORT`?

**Review by mkulke — COMMENTED** [review]
the az-tdx-vtpm verifier has been merged since, should we adjust it in the same PR?

regarding the switch to v0.5: this changed the body of the evidence. we might want to consider versioning the evidence that we produce on the attester side. if someone sends evidence with older guest-components to this verifier it'll fail to serialize, with a potentially confusing error. :/

if we'd send `{ version: "v1" , ... }` from guest-components, we could verify this here with a readable error msg: "Evidence version < v1" or something

Alternatively we can change the Evidence type here to have `pcrs: Option<...>` and fail with a similar error if it's None. but I find the first option preferable

File: attestation-service/verifier/src/az_snp_vtpm/mod.rs (hunk 1 above)
Comment (surajssd) [reply attestation-service/verifier/src/az_snp_vtpm/mod.rs]: Yep fixed this.

**Comment by surajssd** [comment]
> the az-tdx-vtpm verifier has been merged since, should we adjust it in the same PR?

What does this entail?

> regarding the switch to v0.5: this changed the body of the evidence. we might want to consider versioning the evidence that we produce on the attester side. if someone sends evidence with older guest-components to this verifier it'll fail to serialize, with a potentially confusing error. :/
> 
> if we'd send `{ version: "v1" , ... }` from guest-components, we could verify this here with a readable error msg: "Evidence version < v1" or something

How do I do this? As in add the version where?

**Comment by mkulke** [comment]
> > the az-tdx-vtpm verifier has been merged since, should we adjust it in the same PR?
> 
> What does this entail?

bumping the az-tdx crate too, and a similar replacement of the evidence fixtures I guess. should be pretty much 1:1 to the az-snp changes

**Comment by mkulke** [comment]
> > regarding the switch to v0.5: this changed the body of the evidence. we might want to consider versioning the evidence that we produce on the attester side. if someone sends evidence with older guest-components to this verifier it'll fail to serialize, with a potentially confusing error. :/
> > if we'd send `{ version: "v1" , ... }` from guest-components, we could verify this here with a readable error msg: "Evidence version < v1" or something
> 
> How do I do this? As in add the version where?

The version would have to be added to the [evidence type](https://github.com/confidential-containers/guest-components/blob/f603d1e2e38a615b29be302b32136b6bd62743ef/attestation-agent/attester/src/az_snp_vtpm/mod.rs#L24-L29) in guest-components and similarly in the verifier module here:

https://github.com/confidential-containers/kbs/blob/3003ced913bf83fa11d3ef753bb621f9cd030ae8/attestation-service/verifier/src/az_snp_vtpm/mod.rs#L25-L30

At the moment the evidence is serialized as json and maybe serde will happily ignore an additional property `quote.pcrs` when deserializing, so nothing breaks immediately.

but we would have to introduce such a field once we actually use the `pcrs` field, so I'd rather do it here. I would suggest this PR back to draft, until a corresponding change in guest-components has been merged.

**Comment by surajssd** [comment]
Once https://github.com/confidential-containers/guest-components/pull/460 merged, I will udpate this PR to use 0.5.1 for az-cvm-vtpm crates and also update the guest components.

**Comment by Xynnn007** [comment]
@surajssd https://github.com/confidential-containers/guest-components/pull/460 is merged

File: attestation-service/verifier/Cargo.toml [hunk 2]
```diff
@@ -4,22 +4,34 @@ version = "0.1.0"
...
-default = [ "all-verifier" ]
-all-verifier = [ "tdx-verifier", "sgx-verifier", "snp-verifier", "az-snp-vtpm-verifier", "az-tdx-vtpm-verifier", "csv-verifier", "cca-verifier" ]
-tdx-verifier = [ "eventlog-rs", "scroll", "sgx-dcap-quoteverify-rs" ]
-sgx-verifier = [ "scroll", "sgx-dcap-quoteverify-rs" ]
-az-snp-vtpm-verifier = [ "az-snp-vtpm", "sev", "snp-verifier" ]
-az-tdx-vtpm-verifier = [ "az-tdx-vtpm", "openssl", "tdx-verifier" ]
-snp-verifier = [ "asn1-rs", "openssl", "sev", "x509-parser" ]
-csv-verifier = [ "openssl", "csv-rs", "codicon" ]
-cca-verifier = [ "ear", "veraison-apiclient" ]
+default = ["all-verifier"]
```
Comment (mkulke) [inline attestation-service/verifier/Cargo.toml:7]: nit: we should avoid formatting the Cargo.toml file. edit w/ nano or something :)

File: attestation-service/verifier/src/az_snp_vtpm/mod.rs [hunk 3]
```diff
@@ -152,12 +152,18 @@ mod tests {
...
-        wrong_report[0x00b0] = 0;
+        wrong_report[0x01a6] = 0;
         let hcl_report = HclReport::new(wrong_report.to_vec()).unwrap();
         let snp_report = hcl_report.try_into().unwrap();
         let vcek = Vcek::from_pem(include_str!("../../test_data/az-snp-vtpm/vcek.pem")).unwrap();
         let vendor_certs = load_milan_cert_chain().as_ref().unwrap();
-        verify_snp_report(&snp_report, &vcek, vendor_certs).unwrap_err();
+        assert_eq!(
+            verify_snp_report(&snp_report, &vcek, vendor_certs)
+                .err()
```
Comment (mkulke) [inline attestation-service/verifier/src/az_snp_vtpm/mod.rs:162]: unwrap_err()

File: kbs/tools/client/Cargo.toml [hunk 4]
```diff
@@ -18,13 +18,20 @@ base64.workspace = true
 clap = { version = "4.0.29", features = ["derive"] }
 env_logger.workspace = true
 jwt-simple = "0.11.4"
-kbs_protocol = { git = "https://github.com/confidential-containers/guest-components.git", rev = "42b7c96", default-features = false }
+kbs_protocol = { git = "https://github.com/confidential-containers/guest-components.git", rev = "fc89b71", default-features = false }
 log.workspace = true
-reqwest = { version = "0.11.18", default-features = false, features = ["cookies", "json"] }
+reqwest = { version = "0.11.18", default-features = false, features = [
```
Comment (mkulke) [inline kbs/tools/client/Cargo.toml:23]: nit: we should avoid formatting the Cargo.toml file. edit w/ nano or something :)

**Review by mkulke — APPROVED** [review]
small nits, otherwise lgtm

File: attestation-service/verifier/Cargo.toml (hunk 2 above)
Comment (surajssd) [reply attestation-service/verifier/Cargo.toml:7]: Apart from the fact that this is unrelated to the PR, is there a reason to not format `Cargo.toml` file? I see that the [documentation](https://github.com/rust-lang/rust/blob/master/src/doc/style-guide/src/cargo.md) does recommend breaking untenably long lines into multi-line block.

Comment (mkulke) [reply attestation-service/verifier/Cargo.toml:7]: Formatting toml is fine per se, IMO. It's problematic when it's not enforced in the CI and different editors disagree on how to format. In this case there will be bigger diffs flapping between different styles and it's hard to find meaningful changes when you have to bisect.

**Comment by mkulke** [comment]
There is a dependency to guest-components in `./kbs/tools/client/Cargo.toml` we might want the rev in this PR too to reflect the changes on the attester side.

**Review by mkulke — COMMENTED** [review]
There is a quote.verify_pcrs() check missing in each verifier.

Also, I don't think there is a pub fn to retrieve the PCRs from the quote yet. We can address this is a later PR when we actually consume the PCRs, as this probably needs a new minor release in the az-*-vtpm crates.

**Comment by surajssd** [comment]
> There is a quote.verify_pcrs() check missing in each verifier.

Why don't we use the generic `verify` function from the library? https://github.com/kinvolk/azure-cvm-tooling/blob/de7ee4881fe7de01a0253d2e9960363f940ed682/az-cvm-vtpm/src/vtpm/verify.rs#L38-L49

Rather than reimplementing them here: https://github.com/confidential-containers/kbs/blob/3003ced913bf83fa11d3ef753bb621f9cd030ae8/attestation-service/verifier/src/az_tdx_vtpm/mod.rs#L76-L94

**Comment by surajssd** [comment]
> There is a quote.verify_pcrs() check missing in each verifier.

@mkulke done.

File: attestation-service/verifier/src/az_snp_vtpm/mod.rs [hunk 5]
```diff
@@ -48,9 +48,10 @@ impl Verifier for AzSnpVtpm {
     /// The following verification steps are performed:
     /// 1. TPM Quote has been signed by AK included in the HCL variable data
     /// 2. Attestation report_data matches TPM Quote nonce
-    /// 3. SNP report's report_data field matches hashed HCL variable data
-    /// 4. SNP Report is genuine
-    /// 5. SNP Report has been issued in VMPL 0
+    /// 3. TPM Quote's PCRs' digest matches the PCR digest from attestation report
```
Comment (mkulke) [inline attestation-service/verifier/src/az_snp_vtpm/mod.rs:51]: there is no PCR digest in the attestation report, the digest ist in the TPM Quote.

```suggestion
    /// 3. TPM PCRs' digest matches the digest in the Quote
```

File: attestation-service/verifier/src/az_snp_vtpm/mod.rs [hunk 6]
```diff
@@ -106,6 +109,14 @@ fn verify_signature(quote: &Quote, hcl_report: &HclReport) -> Result<()> {
     Ok(())
 }
 
+fn verify_pcrs(quote: &Quote) -> Result<()> {
+    quote.verify_pcrs().context(
+        "Digest of vTPM PCRs from quote do not match the PCR digest from attestation report",
```
Comment (mkulke) [inline attestation-service/verifier/src/az_snp_vtpm/mod.rs:114]:
```suggestion
        "Digest of PCRs does not match digest in Quote",
```

File: attestation-service/verifier/src/az_tdx_vtpm/mod.rs [hunk 7]
```diff
@@ -84,6 +87,14 @@ fn verify_tpm_signature(quote: &TpmQuote, hcl_report: &HclReport) -> Result<()>
     Ok(())
 }
 
+fn verify_pcrs(quote: &TpmQuote) -> Result<()> {
+    quote.verify_pcrs().context(
+        "Digest of vTPM PCRs from quote do not match the PCR digest from attestation report",
```
Comment (mkulke) [inline attestation-service/verifier/src/az_tdx_vtpm/mod.rs:92]:
```suggestion
        "Digest of PCRs does not match digest in Quote",
```

File: attestation-service/verifier/src/az_tdx_vtpm/mod.rs [hunk 8]
```diff
@@ -30,8 +30,9 @@ impl Verifier for AzTdxVtpm {
     /// The following verification steps are performed:
     /// 1. TPM Quote has been signed by AK included in the HCL variable data
     /// 2. Attestation nonce matches TPM Quote nonce
-    /// 3. TD Quote is genuine
-    /// 4. TD Report's report_data field matches hashed HCL variable data
+    /// 3. TPM Quote's PCRs' digest matches the PCR digest from attestation report
```
Comment (mkulke) [inline attestation-service/verifier/src/az_tdx_vtpm/mod.rs:33]:
```suggestion
    /// 3. TPM PCRs' digest matches the digest in the Quote
```

**Review by mkulke — APPROVED** [review]
lgtm

**Comment by surajssd** [comment]
@fitzthum if the code looks good to you can we merge this?

**Pull request merged**